src/SignalR/common/Protocols.Json/src/Protocol/JsonHubProtocol.cs

```diff
@@ -0,0 +1,758 @@
+                        {
+                            try
+                            {
+                                var returnType = binder.GetReturnType(invocationId);
+                                result = BindType(resultToken.RootElement, returnType);
+                            }
+                            finally
+                            {
+                                resultToken.Dispose();
+                            }
+                        }
+
+                        message = BindCompletionMessage(invocationId, error, result, hasResult, binder);
+                        break;
+                    case HubProtocolConstants.CancelInvocationMessageType:
+                        message = BindCancelInvocationMessage(invocationId);
+                        break;
+                    case HubProtocolConstants.PingMessageType:
+                        return PingMessage.Instance;
+                    case HubProtocolConstants.CloseMessageType:
+                        return BindCloseMessage(error);
+                    case null:
+                        throw new InvalidDataException($"Missing required property '{TypePropertyName}'.");
+                    default:
+                        // Future protocol changes can add message types, old clients can ignore them
+                        return null;
+                }
+
+                return ApplyHeaders(message, headers);
+            }
+            catch (JsonReaderException jrex)
+            {
+                throw new InvalidDataException("Error reading JSON.", jrex);
+            }
+        }
+
+        private Dictionary<string, string> ReadHeaders(ref Utf8JsonReader reader)
+        {
+            var headers = new Dictionary<string, string>(StringComparer.Ordinal);
+
+            if (reader.TokenType != JsonTokenType.StartObject)
+            {
+                throw new InvalidDataException($"Expected '{HeadersPropertyName}' to be of type {JsonTokenType.StartObject}.");
+            }
+
+            while (reader.Read())
+            {
+                switch (reader.TokenType)
+                {
+                    case JsonTokenType.PropertyName:
+                        var propertyName = reader.GetString();
+
+                        reader.CheckRead();
+
+                        if (reader.TokenType != JsonTokenType.String)
+                        {
+                            throw new InvalidDataException($"Expected header '{propertyName}' to be of type {JsonTokenType.String}.");
+                        }
+
+                        headers[propertyName] = reader.GetString();
+                        break;
+                    case JsonTokenType.Comment:
+                        break;
+                    case JsonTokenType.EndObject:
+                        return headers;
+                }
+            }
+
+            throw new JsonReaderException("Unexpected end when reading message headers", 0, 0);
```

Comment:
`0, 0` refers to line and position?

Why not throw InvalidDataException here like CheckRead does?